src/sentry/tasks/releasemonitor.py

```diff
@@ -28,51 +28,47 @@ def monitor_release_adoption(**kwargs):
-        #     aggregated_projects = defaultdict(list)
-        #     start_time = time.time()
-        #     offset = 0
-        #     while (time.time() - start_time) < MAX_SECONDS:
-        #         query = Query(
-        #             dataset="sessions",
-        #             match=Entity("org_sessions"),
-        #             select=[
-        #                 Column("org_id"),
-        #                 Column("project_id"),
-        #             ],
-        #             groupby=[Column("org_id"),Column("project_id")],
-        #             where=[
-        #                 Condition(
-        #                     Column("started"), Op.GTE, datetime.utcnow() - timedelta(hours=6)
-        #                 ),
-        #                 Condition(Column("started"), Op.LT, datetime.utcnow()),
-        #             ],
-        #             granularity=Granularity(3600),
-        #             orderby=[OrderBy(Column("org_id"), Direction.ASC)],
-        #         ).set_limit(CHUNK_SIZE+1).set_offset(offset)
-        #         data = snuba.raw_snql_query(query, referrer="tasks.monitor_release_adoption")["data"]
-        #         count = len(data)
-        #         more_results = count >= CHUNK_SIZE
-        #         offset += count
-
-        #         for row in data:
-        #             aggregated_projects[row['org_id']].append(row['project_id'])
-
-        #         if not more_results:
-        #             break
-
-        #     else:
-        #         logger.info(
-        #             "monitor_release_adoption.loop_timeout",
-        #             extra={"offset": offset},
-        #         )
-        # NOTE: Hardcoded data for sentry org and sentry project for early release, in the same format snuba should return
-        aggregated_projects = {
-            1: [1],
-            315582: [5245995],
-            307710: [5227327],
-            34825: [1769835],
-        }  # org id: [project ids]
+        aggregated_projects = defaultdict(list)
+        start_time = time.time()
+        offset = 0
+        while (time.time() - start_time) < MAX_SECONDS:
+            query = (
+                Query(
+                    dataset="sessions",
+                    match=Entity("org_sessions"),
+                    select=[
+                        Column("org_id"),
+                        Column("project_id"),
+                    ],
+                    groupby=[Column("org_id"), Column("project_id")],
+                    where=[
+                        Condition(
+                            Column("started"), Op.GTE, datetime.utcnow() - timedelta(hours=6)
+                        ),
+                        Condition(Column("started"), Op.LT, datetime.utcnow()),
+                    ],
+                    granularity=Granularity(3600),
+                    orderby=[OrderBy(Column("org_id"), Direction.ASC)],
+                )
+                .set_limit(CHUNK_SIZE + 1)
+                .set_offset(offset)
```

Comment:
I don't know if there's any efficiency gain here, but rather than using offset we could have a filter on `org_id, project_id` that we use to page on, similar to how we paginate in the API. I know that using offsets in postgres is inefficient, but not sure if it's the same in Clickhouse.

Reply:
Aah, yeah, like a cursor built on that tuple...I think this should be possible and not too tricky... Maybe @evanh has some knowledge on offsets in clickhouse to see if this is worth pursuing.

Reply:
Can you elaborate on this more? The reason the limit/offset is necessary is because you can only ever fetch at most 10,000 rows from Snuba (we have a hard cap on rows returned). So if there's chance you'll get more rows than that in your result, you have to paginate. If the org/project filter prevents that then yeah you don't need the offset/limit.

Reply:
So I think the idea here would still use limit, but instead of using offset, it would filter by org_id and project_id in such a way that we "page forward". We do this using the cursors in the API (the cursor value can be a date, and we filter based on date (gt or lt depending on sort order), as opposed to the cursor value being a "page" value (0, 1, 2).

Reply:
I guess based on what Dan is saying this is also more performant than just using a page value with offsets in Postgres. I actually mostly thought we did it to improve paging consistency (as the items you're paging through change)

Reply:
@evanh I know that in postgres offset is inefficient, because it fetches the dataset and then has to go through `OFFSET` rows to return the rows you want. So if there are indexed cols it's better to use them to paginate. I'm not sure whether this is the case in Clickhouse though

Reply:
I suspect Clickhouse is similar, as long as the columns you're using are sorting keys (which they are in this case). Is the idea to have conditions like `project_id > 1 and org_id > 2`?

Reply:
Yeah, that was my general thought here

Reply:
Got it. Makes sense.

Reply:
Assuming it's all right with both of you, I'm gonna keep this in mind and possibly follow-up with it as an optimization if needed. I think it might have some hidden complexities (or I'm dumb). Since project id's are non-linear across orgs I can't just do `>` on both, I need the project_id filter to be based on the org_id filter. Else I might miss projects. Also `>` on an org could mean we miss the rest of its projects (so we maybe want >= on org with the project filter, but when we finish the org and move on to the next org, the project filter could cause us to miss projects from that org)

Reply:
We could still miss projects using the current method - pagination could cut off in the middle of an orgs projects. I assume that's fine and it just an optimization?